### PR TITLE
add seq/par and hook

### LINKED ISSUE #43: Add something like setUp and tearDown
Add something like setUp and tearDown, something like hooks in hspec:
https://hspec.github.io/writing-specs.html

### LINKED ISSUE #44: Add the ability to set up variables on the describe block
If you have some state you want to test in multiple `it` blocks, instead of having to instantiate it multiple times, it would be nice to be able to do it once in the `describe` block for those specs. Ie:

```purescript
describe "some state" do
  someRef <- readSTRef myRef
  someArray <- freeze mySTArray

  it "should have this property" do
    ...

  it "should have that property" do
    ...
```

### LINKED ISSUE #63: Custom compile warning for `itOnly` and `describeOnly`
What do you think of the idea to have `itOnly` and `describeOnly` cause compile warnings similar to how `traceAnyA` and friends does. I can't speak for others, but I have accidentally committed an `itOnly` block far too many times, not realizing.

### LINKED ISSUE #68: Let Assertions Return Either
I think a structure like this would provide a much better separation of concerns:

```purescript
shouldEqual :: String -> String -> Either String String
shouldEqual actual expected =
  if (actual /= expected)
  then Left "\"" <> actual <> "\" /= \"" <> expected <> "\""
  else Right ""


shouldBeOk :: forall r. Either String String -> Aff r Unit
shouldBeOk value = case value of
  Left error -> fail error
  Right _ -> (pure unit)


it "should do something" do
  shouldBeOk do
    actual <- getSomeEither
    expected <- getAnotherEither
    actual `shouldEqual` expected
```

Then all assertions would simply be of type `String -> String -> Either String String`

### LINKED ISSUE #80: Add a way to run tests in parallel
I was thinking on adding `describePar` and `describeParOnly`. so Description is marked as parallel and during execution in `_run` we are run producers in parallel. it would be very useful in case ps-spec is used for running integration tests and some tests don't effect each other.

so if you have:
``` purs
describe "delay" do
  it "can delay for 100ms" do
    delay $ Milliseconds $ 100.0 * 1.0
  it "can delay for 200ms" do
    delay $ Milliseconds $ 100.0 * 2.0
  it "can delay for 300ms" do
    delay $ Milliseconds $ 100.0 * 3.0
  it "can delay for 400ms" do
    delay $ Milliseconds $ 100.0 * 4.0
  it "can delay for 500ms" do
    delay $ Milliseconds $ 100.0 * 5.0
```
which takes 1500 ms, you can apply  this change:
``` diff
-describe "Foo" do
+describePar "Foo" do
```
and it will take 500ms.

Main issue would be logging from tests which are executing in parallel, as you would have no way to know from which one a logged message is coming from.

### PR DESCRIPTION
**Pull request by safareli**
fix #80 
fix #43 
fix #44 (any m can be lifted into Spec so you can lift Effect too)
fix #63 
fix #68  (in the end user can hoist Either to Aff)

it might somewhat fixes #73 haven't looked into it yet

---

Initially I tired to rebase #52  on top of #82 but using the approach taken in #52 it was not possible to define `afterAll` and nesting of hooks was not possible too. So instead I ported this version from hspec.

things left to do:
- [x] need to fix reporter to handle case when there is nesting of describe in parallel (see logs of delaySpecExample)
- [x] add docs of hooks (probably copy and modify from hspec)

### CODE REVIEW

**Comment by safareli** [comment]
This is ready for review @felixmulder @owickstrom. tomorrow will try to fix issue with reporter (about parallel nested describes) and add some docs. let me know if there are some comments/suggestions.

**Comment by felixmulder** [comment]
Cool! Won't have time to review today - but hopefully during the weekend 🙂

**Comment by safareli** [comment]
The issue I have now is that `consoleReporter` and `specReporter` need to handle case like this:

```purs
describe "a" $ parallel do
  it "a.1" $ delay $ Milliseconds 500.0
  it "a.2" $ delay $ Milliseconds 500.0
  describe "b" do
    it "b.1" $ delay $ Milliseconds 500.0
    it "b.2" $ delay $ Milliseconds 500.0
```
so a reporter will get events in this order, which is messing output:
```
Suite a
Test a.1
Test a.2
Suite a.b
Test a.b.1
Test a.b.2
Pass a.1 + TestEnd a.1
Pass a.2 + TestEnd a.2
Pass a.b.1 + TestEnd a.b.1
Pass a.b.2 + TestEnd a.b.2
SuiteEnd a.b
SuiteEnd a
```


solution is to make reporters more "advanced", by showing currently running tests with some spinner, and then once they finish update corresponding indicators, ideally so that if a test is logging something to console, it's not interfering to the test output (I haven't done such things in terminal and some suggestions/helpful links, would be great help for fixing this issue)

**Comment by safareli** [comment]
AFAICT without some kind of mocking/monkey patching, a reporter has no way to know if a running test logged something to terminal. If that's true, we have 2 options:
1. ignore case when some test logs to console and be ok that it will mess output.
2. fancy reporters like `consoleReporter` and `specReporter` will take `Bus (Path, String)` as argument and they will be watching for messages and safely printing them so output is not messed up.

I would at first go with option 1 and then probably in new PR implement option 2 so, before that using console.log in `parallel` tests will mess output at bit.

**Comment by safareli** [comment]
so i was able to update spec/console reporters to handle parallel execution:

screen recording 2019-01-15 at 12 22 21 pm
screen recording 2019-01-15 at 12 21 43 pm 1


# But,

if a line printed to console is wrapped, then redraw is messed up and output looks like this:

screen shot 2019-01-15 at 12 33 00 pm
screen shot 2019-01-15 at 12 32 18 pm

this particular issue can be fixed in `countLines` function, if reporter subscribes to width of terminal (with `process.stdout.getWindowSize` and `process.stdout.on("resize")`) but I think it might not work for all terminals&OS combinations.

so at this point I feel that the "redrawing" terminal output is not a good idea, so will try to simplify reporters and will log things to console which are not going to change, it would mean that if you are running multiple tests in parallel you wouldn't see there output until whole all of them are finished.

**Comment by safareli** [comment]
It's working now and is ready for review:
screen recording 2019-01-15 at 6 03 41 pm
screen recording 2019-01-15 at 6 04 33 pm

**Comment by felixmulder** [comment]
Wow, awesome work @safareli !

You've really put time into this.

**Comment by felixmulder** [comment]
Quite a large PR - but I couldn't see anything that raised a red flag with me. @owickstrom, maybe you'd like to review as well?

**Comment by safareli** [comment]
> You've really put time into this.

Yeah, was working on it for about a week, we are using this lib at @f-o-a-m, and we wanted to have this features, so it was sort of my "task".

**Comment by safareli** [comment]
It would be great if someone could check changes made in documentation.

**Comment by owickstrom** [comment]
I'll have a look when I'm back from vacation. Thanks for your effort on this @safareli!

File: test/Test/Spec/RunnerSpec.purs [hunk 1]
```diff
@@ -1,35 +1,58 @@
...
+          runSpecFocused itOnlyTest `shouldEqual`
+            [ Node (Left "a")
+              [ Node (Left "b") [ Leaf "works" $ Just true ]
+              ]
+            ]
         it "supports async" do
           res <- delay (Milliseconds 10.0) *> pure 1
           res `shouldEqual` 1
+  where
+    runSpecFocused t = discardUnfocused (execWriter $ un SpecT t) <#> bimap (const unit) (un Item >>> _.isFocused)
```
Comment (owickstrom) [inline test/Test/Spec/RunnerSpec.purs]: `runSpecFocused`, if I'm not misunderstanding, isn't running the full `runSpec` machinery, like this test did before. It seems it's collecting a tree structure of `Maybe Bool` values for `isFocused`. Would it be possible to rewrite this to use runSpec (or some underlying function) to exercise the runner module?

Comment (safareli) [reply test/Test/Spec/RunnerSpec.purs]: Most I could do is to add `collect` let me know if there is something you think is worth adding/changing.

**Comment by safareli** [comment]
Besides what was commented as part of the review I have change type of `hoistSpec` and `mapSpecTree` so the type in which `SpecTree`s are accumulated can also be changed, it can be useful for example if some part of test is using `Spec` type (which uses `Identity`) and then you want to use it in some `SpecT` with other monad like this: `hoistSpec (un Identity >>> pure) (const identity) someTest`

**Comment by safareli** [comment]
After we merge this we should probably update ps-spec-* libs

- `purescript-spec-discovery` because now one can interleave some monadic effect in test generation the `discover` can have signature like this: `discover :: MoandEffect m => String -> SpecT g i m Unit`.
- `purescript-spec-reporter-xunit` should be easy to update.
- `purescript-spec-quickcheck` doesn't even depend on ps-spec and I guess needs no change.
- `purescript-spec-mocha` updating this might not be trivial as we support parallel execution of tests but mocha doesn't. as I understand the main (if not only) benefit of using mocha for user is ability to run tests both in browser and node. it can be accomplished by striping suing console.log instead of process.stdout.write (this [article](https://medium.freecodecamp.org/how-to-get-the-most-out-of-the-javascript-console-b57ca9db3e6d) can be useful for using colors etc in browser console output). so what i propose is to deprecate that and add make our built in reporters work in browser.

Let me know what others think about this.

**Comment by safareli** [comment]
Also do you think it's worth releasing this changes as `v4.0.0-rc1` (maybe people can play with it and get some feedback) or we should go straight to v4?

**Comment by safareli** [comment]
🏓 ping

**Comment by owickstrom** [comment]
I'm good with these last changes! And yeah, I think releasing an RC would be good.

**Comment by owickstrom** [comment]
@felixmulder you good with merging this?

**Comment by owickstrom** [comment]
👏

**Comment by safareli** [comment]
Made a release [v4.0.0-rc1](https://github.com/purescript-spec/purescript-spec/releases/tag/v4.0.0-rc1)

**Pull request merged**